cache/cache.go

```diff
@@ -0,0 +1,193 @@
+package cache
+
+import (
+	"bytes"
+	"sync"
+
+	"github.com/google/btree"
+	"go.uber.org/atomic"
+	"golang.org/x/exp/slices"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+)
+
+type ATXData struct {
+	Weight             uint64
+	BaseHeight, Height uint64
+	Nonce              types.VRFPostIndex
+	Malicious          bool
+}
+
+type Opt func(*Cache)
+
+// WithCapacity sets the number of epochs from the latest applied
+// that cache will maintain in memory.
+func WithCapacity(capacity types.EpochID) Opt {
+	return func(cache *Cache) {
+		cache.capacity = capacity
+	}
+}
+
+func New(opts ...Opt) *Cache {
+	cache := &Cache{
+		capacity: 2,
+		epochs:   map[types.EpochID]epochCache{},
+	}
+	for _, opt := range opts {
+		opt(cache)
+	}
+	return cache
+}
+
+type Cache struct {
+	evicted atomic.Uint32
+
+	// number of epochs to keep
+	// capacity is not enforced by the cache itself
+	capacity types.EpochID
+
+	mu     sync.RWMutex
+	epochs map[types.EpochID]epochCache
+}
+
+type stored struct {
+	node types.NodeID
+	atx  types.ATXID
+	data *ATXData
```

Comment:
thats a leftover. i was experimenting with storing `stored` as non-pointer, i thought that it will remove pointer chasing during btree traversal and reduce cpu cache misses. but i also suspect that memory in benchmark is allocated not in the best way to highlight problems